File: georocket-server/src/test/java/io/georocket/output/MultiMergerTest.java
package io.georocket.output;

import java.util.Arrays;

import org.apache.commons.lang3.tuple.Pair;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import io.georocket.storage.ChunkMeta;
import io.georocket.storage.ChunkReadStream;
import io.georocket.storage.GeoJsonChunkMeta;
import io.georocket.storage.XMLChunkMeta;
import io.georocket.util.XMLStartElement;
import io.georocket.util.io.BufferWriteStream;
import io.georocket.util.io.DelegateChunkReadStream;
import io.vertx.core.buffer.Buffer;
import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.ext.unit.junit.RunTestOnContext;
import io.vertx.ext.unit.junit.VertxUnitRunner;
import rx.Observable;

/**
 * Tests for {@link MultiMerger}
 * @author Michel Kraemer
 */
@RunWith(VertxUnitRunner.class)
public class MultiMergerTest {
  private static final String XMLHEADER = "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"yes\"?>\n";
  
  /**
   * Run the test on a Vert.x test context
   */
  @Rule
  public RunTestOnContext rule = new RunTestOnContext();
  
  private void doMerge(TestContext context, Observable<Buffer> chunks,
      Observable<ChunkMeta> metas, String jsonContents) {
    MultiMerger m = new MultiMerger(false);
    BufferWriteStream bws = new BufferWriteStream();
    Async async = context.async();
    metas
      .flatMapSingle(meta -> m.init(meta).toSingleDefault(meta))
      .toList()
      .flatMap(l -> chunks.map(DelegateChunkReadStream::new)
          .<ChunkMeta, Pair<ChunkReadStream, ChunkMeta>>zipWith(l, Pair::of))
      .flatMapCompletable(p -> m.merge(p.getLeft(), p.getRight(), bws))
      .toCompletable()
      .subscribe(() -> {
        m.finish(bws);
        context.assertEquals(jsonContents, bws.getBuffer().toString("utf-8"));
        async.complete();
      }, context::fail);
  }
  
  /**
   * Test if one GeoJSON geometry is rendered directly
   * @param context the Vert.x test context
   */
  @Test
  public void geoJsonOneGeometry(TestContext context) {
    String strChunk1 = "{\"type\":\"Polygon\"}";
    Buffer chunk1 = Buffer.buffer(strChunk1);
    GeoJsonChunkMeta cm1 = new GeoJsonChunkMeta("Polygon", "geometries", 0, chunk1.length());
    doMerge(context, Observable.just(chunk1), Observable.just(cm1), strChunk1);
  }
  
  /**
   * Test if two GeoJSON features can be merged to a feature collection
   * @param context the Vert.x test context
   */
  @Test
  public void geoJsonTwoFeatures(TestContext context) {
    String strChunk1 = "{\"type\":\"Feature\"}";
    String strChunk2 = "{\"type\":\"Feature\",\"properties\":{}}";
    Buffer chunk1 = Buffer.buffer(strChunk1);
    Buffer chunk2 = Buffer.buffer(strChunk2);
    GeoJsonChunkMeta cm1 = new GeoJsonChunkMeta("Feature", "features", 0, chunk1.length());
    GeoJsonChunkMeta cm2 = new GeoJsonChunkMeta("Feature", "features", 0, chunk2.length());
    doMerge(context, Observable.just(chunk1, chunk2), Observable.just(cm1, cm2),
      "{\"type\":\"FeatureCollection\",\"features\":[" + strChunk1 + "," + strChunk2 + "]}");
  }
  
  /**
   * Test if simple XML chunks can be merged
   * @param context the Vert.x test context
   */
  @Test
  public void xmlSimple(TestContext context) {
    Buffer chunk1 = Buffer.buffer(XMLHEADER + "<root><test chunk=\"1\"></test></root>");
    Buffer chunk2 = Buffer.buffer(XMLHEADER + "<root><test chunk=\"2\"></test></root>");
    XMLChunkMeta cm = new XMLChunkMeta(Arrays.asList(new XMLStartElement("root")),
        XMLHEADER.length() + 6, chunk1.length() - 7);
    doMerge(context, Observable.just(chunk1, chunk2), Observable.just(cm, cm),
      XMLHEADER + "<root><test chunk=\"1\"></test><test chunk=\"2\"></test></root>");
  }
  
  /**
   * Test if the merger fails if chunks with a different type should be merged
   * @param context the Vert.x test context
   */
  @Test
  public void mixedInit(TestContext context) {
    String strChunk1 = "{\"type\":\"Feature\"}";
    String strChunk2 = XMLHEADER + "<root><test chunk=\"2\"></test></root>";
    Buffer chunk1 = Buffer.buffer(strChunk1);
    Buffer chunk2 = Buffer.buffer(strChunk2);
    
    GeoJsonChunkMeta cm1 = new GeoJsonChunkMeta("Feature", "features", 0, chunk1.length());
    XMLChunkMeta cm2 = new XMLChunkMeta(Arrays.asList(new XMLStartElement("root")),
      XMLHEADER.length() + 6, chunk2.length() - 7);
    
    MultiMerger m = new MultiMerger(false);
    Async async = context.async();
    m.init(cm1)
      .andThen(m.init(cm2))
      .subscribe(context::fail, err -> {
        context.assertTrue(err instanceof IllegalStateException);
        async.complete();
      });
  }
  
  /**
   * Test if the merger fails if chunks with a different type should be merged
   * @param context the Vert.x test context
   */
  @Test
  public void mixedMerge(TestContext context) {
    String strChunk1 = "{\"type\":\"Feature\"}";
    String strChunk2 = XMLHEADER + "<root><test chunk=\"2\"></test></root>";
    Buffer chunk1 = Buffer.buffer(strChunk1);
    Buffer chunk2 = Buffer.buffer(strChunk2);
    
    GeoJsonChunkMeta cm1 = new GeoJsonChunkMeta("Feature", "features", 0, chunk1.length());
    XMLChunkMeta cm2 = new XMLChunkMeta(Arrays.asList(new XMLStartElement("root")),
      XMLHEADER.length() + 6, chunk2.length() - 7);
    
    MultiMerger m = new MultiMerger(false);
    BufferWriteStream bws = new BufferWriteStream();
    Async async = context.async();
    m.init(cm1)
      .andThen(m.merge(new DelegateChunkReadStream(chunk1), cm1, bws))
      .andThen(m.merge(new DelegateChunkReadStream(chunk2), cm2, bws))
      .subscribe(context::fail, err -> {
        context.assertTrue(err instanceof IllegalStateException);
        async.complete();
      });
  }
}

File: georocket-server/src/test/kotlin/io/georocket/output/MultiMergerTest.kt
package io.georocket.output

import io.georocket.assertThatThrownBy
import io.georocket.coVerify
import io.georocket.storage.ChunkMeta
import io.georocket.storage.GeoJsonChunkMeta
import io.georocket.storage.XMLChunkMeta
import io.georocket.util.XMLStartElement
import io.georocket.util.io.BufferWriteStream
import io.georocket.util.io.DelegateChunkReadStream
import io.vertx.core.Vertx
import io.vertx.core.buffer.Buffer
import io.vertx.junit5.VertxExtension
import io.vertx.junit5.VertxTestContext
import io.vertx.kotlin.coroutines.dispatcher
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith

/**
 * Tests for [MultiMerger]
 * @author Michel Kraemer
 */
@ExtendWith(VertxExtension::class)
class MultiMergerTest {
  companion object {
    private const val XMLHEADER = """<?xml version="1.0" encoding="UTF-8" standalone="yes"?>${"\n"}"""
  }

  private fun doMerge(vertx: Vertx, ctx: VertxTestContext, chunks: List<Buffer>,
      metas: List<ChunkMeta>, contents: String) {
    val m = MultiMerger(false)
    val bws = BufferWriteStream()

    for (meta in metas) {
      m.init(meta)
    }

    GlobalScope.launch(vertx.dispatcher()) {
      ctx.coVerify {
        for ((meta, chunk) in metas.zip(chunks)) {
          val stream = DelegateChunkReadStream(chunk)
          m.merge(stream, meta, bws)
        }
        m.finish(bws)
        assertThat(contents).isEqualTo(bws.buffer.toString("utf-8"))
      }
      ctx.completeNow()
    }
  }

  /**
   * Test if one GeoJSON geometry is rendered directly
   */
  @Test
  fun geoJsonOneGeometry(vertx: Vertx, ctx: VertxTestContext) {
    val strChunk1 = """{"type":"Polygon"}"""
    val chunk1 = Buffer.buffer(strChunk1)
    val cm1 = GeoJsonChunkMeta("Polygon", "geometries", 0, chunk1.length())

    doMerge(vertx, ctx, listOf(chunk1), listOf(cm1), strChunk1)
  }

  /**
   * Test if two GeoJSON features can be merged to a feature collection
   */
  @Test
  fun geoJsonTwoFeatures(vertx: Vertx, ctx: VertxTestContext) {
    val strChunk1 = """{"type":"Feature"}"""
    val strChunk2 = """{"type":"Feature","properties":{}}"""

    val chunk1 = Buffer.buffer(strChunk1)
    val chunk2 = Buffer.buffer(strChunk2)

    val cm1 = GeoJsonChunkMeta("Feature", "features", 0, chunk1.length())
    val cm2 = GeoJsonChunkMeta("Feature", "features", 0, chunk2.length())

    doMerge(vertx, ctx, listOf(chunk1, chunk2), listOf(cm1, cm2),
        """{"type":"FeatureCollection","features":[$strChunk1,$strChunk2]}""")
  }

  /**
   * Test if simple XML chunks can be merged
   */
  @Test
  fun xmlSimple(vertx: Vertx, ctx: VertxTestContext) {
    val chunk1 = Buffer.buffer("""$XMLHEADER<root><test chunk="1"></test></root>""")
    val chunk2 = Buffer.buffer("""$XMLHEADER<root><test chunk="2"></test></root>""")

    val cm = XMLChunkMeta(listOf(XMLStartElement("root")),
        XMLHEADER.length + 6, chunk1.length() - 7)

    doMerge(vertx, ctx, listOf(chunk1, chunk2), listOf(cm, cm),
        """$XMLHEADER<root><test chunk="1"></test><test chunk="2"></test></root>""")
  }

  /**
   * Test if the merger fails if chunks with a different type should be merged
   */
  @Test
  fun mixedInit(vertx: Vertx, ctx: VertxTestContext) {
    val strChunk1 = """{"type":"Feature"}"""
    val strChunk2 = """$XMLHEADER<root><test chunk="2"></test></root>"""

    val chunk1 = Buffer.buffer(strChunk1)
    val chunk2 = Buffer.buffer(strChunk2)

    val cm1 = GeoJsonChunkMeta("Feature", "features", 0, chunk1.length())
    val cm2 = XMLChunkMeta(listOf(XMLStartElement("root")),
        XMLHEADER.length + 6, chunk2.length() - 7)

    val m = MultiMerger(false)

    GlobalScope.launch(vertx.dispatcher()) {
      ctx.coVerify {
        m.init(cm1)
        assertThatThrownBy {
          m.init(cm2)
        }.isInstanceOf(IllegalStateException::class.java)
      }
      ctx.completeNow()
    }
  }

  /**
   * Test if the merger fails if chunks with a different type should be merged
   */
  @Test
  fun mixedMerge(vertx: Vertx, ctx: VertxTestContext) {
    val strChunk1 = """{"type":"Feature"}"""
    val strChunk2 = """$XMLHEADER<root><test chunk="2"></test></root>"""

    val chunk1 = Buffer.buffer(strChunk1)
    val chunk2 = Buffer.buffer(strChunk2)

    val cm1 = GeoJsonChunkMeta("Feature", "features", 0, chunk1.length())
    val cm2 = XMLChunkMeta(listOf(XMLStartElement("root")),
        XMLHEADER.length + 6, chunk2.length() - 7)

    val m = MultiMerger(false)
    val bws = BufferWriteStream()

    GlobalScope.launch(vertx.dispatcher()) {
      ctx.coVerify {
        m.init(cm1)
        m.merge(DelegateChunkReadStream(chunk1), cm1, bws)
        assertThatThrownBy {
          m.merge(DelegateChunkReadStream(chunk2), cm2, bws)
        }.isInstanceOf(IllegalStateException::class.java)
      }
      ctx.completeNow()
    }
  }
}
